worlds/kh2/test/test_form_logic.py

```diff
@@ -0,0 +1,217 @@
+            allFormsCopy = self.allForms.copy()
+            allFormsCopy.remove(form)
+            self.collect(self.get_item_by_name(form))
+            for _ in range(self.count(ItemName.LightDarkness)):
+                self.remove(self.get_item_by_name(ItemName.LightDarkness))
+            self.assertTrue((self.can_reach_location(levels[0])), levels[0])
+            self.assertTrue((self.can_reach_location(levels[1])), levels[1])
+            self.assertFalse((self.can_reach_location(levels[2])), levels[2])
+            for i in range(3):
+                self.collect(self.get_item_by_name(allFormsCopy[i]))
+                # for some reason after collecting a form it can pick up light and darkness
+                for _ in range(self.count(ItemName.LightDarkness)):
+                    self.remove(self.get_item_by_name(ItemName.LightDarkness))
+
+                self.assertTrue((self.can_reach_location(levels[2 + i])))
+                if i < 2:
+                    self.assertFalse((self.can_reach_location(levels[3 + i])))
+                else:
+                    self.collect(self.get_item_by_name(allFormsCopy[i + 1]))
+                    for _ in range(self.count(ItemName.LightDarkness)):
+                        self.remove(self.get_item_by_name(ItemName.LightDarkness))
+                    self.assertTrue((self.can_reach_location(levels[3 + i])))
+
+    def test_default_with_lnd(self):
+        allPossibleForms = self.allForms
+        self.collect_all_but(allPossibleForms)
+        for form, levels in self.driveFormMap.items():
+            if form != ItemName.FinalForm:
+                for driveForm in self.allForms:
+                    for _ in range(self.count(driveForm)):
+                        self.remove(self.get_item_by_name(driveForm))
+                allFormsCopy = self.allForms.copy()
+                allFormsCopy.remove(form)
+                self.collect(self.get_item_by_name(ItemName.LightDarkness))
+                self.assertFalse((self.can_reach_location(levels[0])))
+                self.collect(self.get_item_by_name(form))
+
+                self.assertTrue((self.can_reach_location(levels[0])))
+                self.assertTrue((self.can_reach_location(levels[1])))
+                self.assertTrue((self.can_reach_location(levels[2])))
+                self.assertFalse((self.can_reach_location(levels[3])))
+                for i in range(2):
+                    self.collect(self.get_item_by_name(allFormsCopy[i]))
+                    self.assertTrue((self.can_reach_location(levels[i + 3])))
+                    if i <= 2:
+                        self.assertFalse((self.can_reach_location(levels[i + 4])))
+
+
+class TestJustAForm(KH2TestFormBase):
+    # this test checks if you can unlock final form with just a form.
+    options = {
+        "AutoFormLogic":  False,
+        "FinalFormLogic": "just_a_form"
+    }
+
+    def test_nothing(self):
```

Comment:
to test that false auto form logic and final form logic with just a form clears the normal tests

Reply:
the normal tests (methods that start with test_) are automatically run on every class that inherits from WorldTestBase. you don't have to do anything to instantiate them because they already exist. https://github.com/ArchipelagoMW/Archipelago/pull/2348/files#diff-364ece66f0a11a275c392299f96afd2f2856e3a3bf4798d76dd1d2f6f6b07587R53

Reply:
ah that makes a lot more sense